File: cpp/543.cpp
#include <algorithm>
#include <climits>
#include <iostream>

using namespace std;

// Definition for a binary tree node.
struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode() : val(0), left(nullptr), right(nullptr) {}
  TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
  TreeNode(int x, TreeNode *left, TreeNode *right)
      : val(x), left(left), right(right) {}
};

class Solution {
public:
  int ret;
  int path(TreeNode *cur) {
    if (cur == nullptr)
      return 0;
    int left = path(cur->left);
    int right = path(cur->right);
    if (ret < left + right)
      ret = left + right;
    return max(left, right) + 1;
  }
  int diameterOfBinaryTree(TreeNode *root) {
    ret = INT_MIN;
    path(root);
    return ret;
  }
};

File: go/543.go
package main

// Definition for a binary tree node.
type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

var ret int

func max(a int, b int) int {
	if a > b {
		return a
	}
	return b
}

func helper(t *TreeNode) int {
	if t == nil {
		return 0
	}
	var l, r int = helper(t.Left), helper(t.Right)
	ret = max(ret, l+r)
	return max(l, r) + 1
}

func diameterOfBinaryTree(root *TreeNode) int {
	ret = -1 << 31
	helper(root)
	return ret
}
